src/shapes/Textbox.ts

```diff
@@ -319,47 +372,35 @@ export class Textbox extends IText {
     _line,
     lineIndex: number,
     desiredWidth: number,
+    WordsWidthData: WordsWidthData,
     reservedSpace = 0
   ): Array<any> {
     const additionalSpace = this._getWidthOfCharSpacing(),
       splitByGrapheme = this.splitByGrapheme,
       graphemeLines = [],
-      words = splitByGrapheme
-        ? this.graphemeSplit(_line)
-        : this.wordSplit(_line),
       infix = splitByGrapheme ? '' : ' ';
 
     let lineWidth = 0,
       line = [],
       // spaces in different languages?
       offset = 0,
       infixWidth = 0,
-      largestWordWidth = 0,
       lineJustStarted = true;
-    // fix a difference between split and graphemeSplit
-    if (words.length === 0) {
-      words.push([]);
-    }
+
     desiredWidth -= reservedSpace;
-    // measure words
-    const data = words.map((word) => {
-      // if using splitByGrapheme words are already in graphemes.
-      word = splitByGrapheme ? word : this.graphemeSplit(word);
-      const width = this._measureWord(word, lineIndex, offset);
-      largestWordWidth = Math.max(width, largestWordWidth);
-      offset += word.length + infix.length;
-      return { word, width };
-    });
+
+    const { largestWordWidth, wordsData } = WordsWidthData;
 
     const maxWidth = Math.max(
       desiredWidth,
       largestWordWidth,
       this.dynamicMinWidth
     );
     // layout words
+    const data = wordsData[lineIndex];
     offset = 0;
     let i;
-    for (i = 0; i < words.length; i++) {
+    for (i = 0; i < data.length; i++) {
```

Comment:
there is probably something else that can be done down here to bring infix width calculation in measure words.
but i m intentionally skipping it for now.
Seems like line 419 could be pused in the else condition after 415.

i m also not convinced at all from the offset++ at line 426, but eventually it works

Reply:
Yes, all this is correct.
I did that in the other PR.

Reply:
Now it will make sense to you.